File: tree/5639/main.cpp
// Authored by : ccocco0609
// Co-authored by : -
// Link : http://boj.kr/681f9d68a83c4f6a8627bd58b3db7178

#include <bits/stdc++.h>
using namespace std;

struct Node {
	friend class BinaryTree;
private:
	Node *left;
	Node *right;
	int key;
public:
	Node(int key, Node *left = NULL, Node *right = NULL) {
		this->key = key;
		this->left = left;
		this->right = right;
	}
};

class BinaryTree {
private:
	Node *root;
public:
	BinaryTree(int key) {
		root = new Node(key);
	}

	Node *getRoot() {
		return root;
	}

	void insert_node(Node *current, int key) {
		if (current->key > key) {
			if (current->left == NULL) {
				current->left = new Node(key);
			}
			else {
				insert_node(current->left, key);
			}
		}
		else if (current->key < key) {
			if (current->right == NULL) {
				current->right = new Node(key);
			}
			else {
				insert_node(current->right, key);
			}
		}
	}

	void print_cur_key(Node *current) {
		cout << current->key << '\n';
	}
	
	void postorder(Node *current) {
		if (current != NULL) {
			postorder(current->left);
			postorder(current->right);
			print_cur_key(current);
		}
	}
};

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	
	int first;
	
	cin >> first;

	BinaryTree tree(first);

	int tmp;

	while (cin >> tmp) {
		tree.insert_node(tree.getRoot(), tmp);
	}

	tree.postorder(tree.getRoot());

	return 0;
}

File: tree/5639/Main.java
//Authored by : vswngjs
//Co-authored by : -
//Link : http://boj.kr/095123c8145c4bb988e87a75cd4fc2f7
import java.util.*;
import java.io.*;
public class Main {
    static class Node {
        int num;
        Node left, right;

        Node(int num) {
            this.num = num;
        }

        Node(int num, Node left, Node right) {
            this.num = num;
            this.left = left;
            this.right = right;
        }

        void insert(int n) {
            if (n < this.num) {
                if (this.left == null) {
                    this.left = new Node(n);
                }
                else {
                    this.left.insert(n);
                }
            }
            else {
                if (this.right == null) {
                    this.right = new Node(n);
                }
                else {
                    this.right.insert(n);
                }
            }
        }
    }
    public static void main(String [] args) throws IOException {
            FastReader br = new FastReader();
            Node root = new Node(Integer.parseInt(br.nextLine()));
            String input;
            while (true) {
                input = br.nextLine();
                if (input == null || input.equals("")) {
                    break;
                }
                root.insert(Integer.parseInt(input));
            }
            postorder(root);


    }
    static void postorder(Node node) {
        if (node == null) {
            return;
        }
        postorder(node.left);
        postorder(node.right);
        System.out.println(node.num);
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while(st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                }
                catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() { return Integer.parseInt(next()); }
        long nextLong() { return Long.parseLong(next()); }
        double nextDouble() { return Double.parseDouble(next()); }
        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}

